src/motion.rs

```diff
@@ -0,0 +1,208 @@
+#![deny(missing_docs)]
+//! Motion.
+
+use std::{fs, path::Path};
+
+use crate::{
+    core::Model,
+    error::CubismResult,
+    json::motion::{Motion3, Segment, SegmentPoint},
+};
+
+trait Interpolate {
+    fn contains(&self, t: f32) -> bool;
+    fn interpolate(&self, t: f32) -> f32;
+}
+
+fn lerp_points(p0: SegmentPoint, p1: SegmentPoint, t: f32) -> SegmentPoint {
+    SegmentPoint {
+        time: p0.time + (p1.time - p0.time) * t,
+        value: p0.value + (p1.value - p0.value) * t,
+    }
+}
+
+impl Interpolate for Segment {
+    fn contains(&self, t: f32) -> bool {
+        match self {
+            Segment::Linear(p0, p1) => p0.time <= t && t < p1.time,
+            Segment::Bezier([p0, _, _, p1]) => p0.time <= t && t < p1.time,
+            Segment::Stepped(p0, t1) => p0.time <= t && t < *t1,
+            Segment::InverseStepped(t0, p1) => *t0 <= t && t < p1.time,
+        }
+    }
+
+    fn interpolate(&self, t: f32) -> f32 {
+        match self {
+            Segment::Linear(p0, p1) => {
+                let k = (t - p0.time) / (p1.time - p0.time);
+
+                if k > 0.0 {
+                    p0.value + (p1.value - p0.value) * k
+                } else {
+                    p0.value
+                }
+            },
+            Segment::Bezier([p0, p1, p2, p3]) => {
+                let k = (t - p0.time) / (p3.time - p0.time);
+                let k = if k < 0.0 { 0.0 } else { k };
+
+                let (p0, p1, p2, p3) = (*p0, *p1, *p2, *p3);
+
+                let p01 = lerp_points(p0, p1, k);
+                let p12 = lerp_points(p1, p2, k);
+                let p23 = lerp_points(p2, p3, k);
+
+                let p012 = lerp_points(p01, p12, k);
+                let p123 = lerp_points(p12, p23, k);
+
+                lerp_points(p012, p123, k).value
+            },
+            Segment::Stepped(p0, _) => p0.value,
+            Segment::InverseStepped(_, p1) => p1.value,
+        }
+    }
+}
+
+/// Handles motions and animates a model.
+#[derive(Clone, Debug)]
+pub struct Motion {
+    json: Motion3,
+    duration: f32,
+    fps: f32,
+    looped: bool,
+    playing: bool,
+    current_time: f64,
+}
+
+impl Motion {
+    /// Creates a Motion from a Motion3.
+    pub fn new(motion3: Motion3) -> Motion {
+        let duration = motion3.meta.duration;
+        let fps = motion3.meta.fps;
+        let looped = motion3.meta.looped;
+
+        Motion {
+            json: motion3,
+            duration,
+            fps,
+            looped,
+            playing: false,
+            current_time: 0.0,
+        }
+    }
+
+    /// Plays a motion.
+    pub fn play(&mut self) {
+        self.playing = true;
+    }
+
+    /// Pauses a motion.
+    pub fn pause(&mut self) {
+        self.playing = false;
+    }
+
+    /// Creates a Motion from a path of .motion3.json file.
+    pub fn from_motion3_json<P: AsRef<Path>>(path: P) -> CubismResult<Motion> {
+        let json = Motion3::from_reader(fs::File::open(path.as_ref())?)?;
```

Comment:
If I am not mistaken you can actually just pass `path` to `File::open` instead of calling `as_ref` on it since `File::open` takes the same generic parameter as this function.